### PR TITLE
Add Isar database to benchmark

### PR DESCRIPTION
**Pull request by peterlazar1993**
@vaind 

This is my first PR to an open source project!!

[Isar](https://github.com/isar/isar) is the in development database from the creator of Hive, [@leisim](https://github.com/leisim).
Even though it is in Alpha, I think it'll be fun to see how it holds up against ObjectBox 1.0
So this PR adds support for benchmarking Isar.

Since Isar expects a zero-argument Constructor for it's model, I had to deviate from the current implementation of Executors.

- All model classes must now implement the [ITestEntity interface](https://github.com/peterlazar1993/objectbox-dart-performance/blob/main/lib/model_interface.dart#L1). [Example](https://github.com/peterlazar1993/objectbox-dart-performance/blob/main/lib/isar_model.dart#L5)
- All Executors must extend from [ExecutorBase](https://github.com/peterlazar1993/objectbox-dart-performance-1/blob/main/lib/executor.dart#L4) and it expects a type argument which implements ITestEntity. [Example](https://github.com/peterlazar1993/objectbox-dart-performance/blob/main/lib/isar_sync_executor.dart#L10)

#### Questions
- Would you be open to removing firebase from the benchmark? Since it's already been [established](https://objectbox.io/objectbox-dart-flutter-v0-11-database-performance-relations/) that firebase is on another league altogether.
- Would it be ok to implement `prepareData()` [method per executor](https://github.com/peterlazar1993/objectbox-dart-performance/blob/main/lib/isar_sync_executor.dart#L21-L34)? `prepareData()` method of the executors generate models with id set as zero. ObjectBox auto-assigns id when it's 0 but almost all of the other DBs require extra processing to set ids in the `insertMany()` method., for eg. [hive](https://github.com/peterlazar1993/objectbox-dart-performance/blob/main/lib/hive_executor.dart#L32-L41)

PS: Feel free to discard this PR; I saw in your recent blog post that updated benchmarks are going to be released soon.
Working on this PR itself was fun cause I got to familiarise with the different DB APIs

### CODE REVIEW

File: lib/isar_sync_executor.dart [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+      );
+    }
+    return result.toList(growable: false);
+  }
+
+  Future<void> close() => _box.close();
+
+  Future<void> insertMany(List<IsarModel> items) => Future.value(tracker.track(
+      'insertMany',
+      () => _box.writeTxnSync((isar) => isar.isarModels.putAllSync(items))));
```
Comment (vaind) [inline lib/isar_sync_executor.dart]: why not `isar.testEntitys.putAllSync()`?

**Review by vaind — CHANGES_REQUESTED** [review]
Thanks for the interest and PR! :+1: 

Looking at the code changes, this PR seems to be trying to do a bit too much... 
It'd be nice if you could extract isar-only as in this PR and leave the rest of the changes as separate PRs, if you'd like.

> Would you be open to removing firebase from the benchmark? Since it's already been established that firebase is on another league altogether.

It's disabled on the current main, so I'm in no rush to remove the code at the moment, just if we want to bring it back in quickly.

> Would it be ok to implement prepareData() method per executor? prepareData() method of the executors generate models with id set as zero. ObjectBox auto-assigns id when it's 0 but almost all of the other DBs require extra processing to set ids in the insertMany() method., for eg. hive

Hmm, this is a dilemma... on one hand, it's nice if all the databases do the same, on the other, some users may choose to tailor their app to a specific DB implementation... Let's see about that (but not in this PR :) )

**Comment by peterlazar1993** [comment]
I have reduced the PR's scope to only add Isar.

The public constructor for [`TestEntity`](https://github.com/objectbox/objectbox-dart-performance/pull/1/files#diff-3139692a101bb10901975db8a91c8ab4db05cc68ccd7c7b614c3c6ca1e12ad82L16) has been removed since Isar requires a zero-arg constructor. The object generation at various places have also been changed to support this.

This PR is not ready to merge since I am running into a crash on iOS simulator and Android emulator. I have raised an [issue](https://github.com/isar/isar/issues/109).
Will update once it is resolved.

File: lib/model.dart [hunk 2]
```diff
@@ -1,29 +1,30 @@
...
+  late int tInt; // 32-bit
 
   @HiveField(3)
-  int tLong; // 64-bit
+  late int tLong; // 64-bit
 
   @HiveField(4)
-  double tDouble;
-
-  TestEntity(this.id, this.tString, this.tInt, this.tLong, this.tDouble);
```
Comment (vaind) [inline lib/model.dart:25]: maybe leave as is and add another public constructor for Isar (only needed until they support `final` fields, which I'm sure they would - that would basically mean it has to recognize constructors).

This should work:
```dart
  // Isar v0.4: Class needs to have a public zero-arg constructor.
  TestEntity.forIsar()
      : id = 0,
        tString = '',
        tInt = 0,
        tLong = 0,
        tDouble = 0;
```

**Review by vaind — COMMENTED** [review]
Look pretty good to me, please consider my other suggestion.

**Comment by vaind** [comment]
and don't forget to mark the PR as "ready" once you consider it done

File: lib/model.dart (hunk 2 above)
Comment (peterlazar1993) [reply lib/model.dart:25]: Yes `final` field support is incoming I believe.
The proposed solution for the constructor _almost_ works. The generator does not pick up the zero arg constructor and instead uses the Class name to generate the adapter. This is a bug of course.

FYI to run this PR, after the build_runner step is done, manually edit the `isar.g.dart` to use the `forIsar` constructor in the `deserialize` method.

I will wait for a while until these issues are resolved.

**Comment by vaind** [comment]
replaced by #3

**Comment by vaind** [comment]
I've sourced the first commits from this PR and created a follow-up for the updated benchmark app

**Pull request closed without merging**